_posts/2022-05-14-appsync-graphql-subscriptions.md

```diff
@@ -0,0 +1,308 @@
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+```
+
+#### Step3:
+
+In your UI component,use `useSubscription` Hook to execute a subscription from React. The following example component uses the subscription we defined earlier to render the most recent order status. Whenever the GraphQL server pushes a new status to the client, the component re-renders with the updated order.
+
+```
+---- LatestOrderStatus.tsx file
+
+import { ApolloError } from 'apollo-client';
+import React from 'react';
+import {  useSubscription } from '@apollo/react-hooks';
+
+const LatestOrderStatus = ({ orderId: string }) =>  {
+  const { data, loading } = useSubscription(
+    GET_ORDER_STATUS,
+    { variables: { orderId },
+    onSubscriptionData: order => {
+      // Add your code here for whatever is going to happen when subscriber gets the new order data.
+    }
+     }
+  );
+  return <h4> You order status is: {!loading && data.getOrderStatus.status}</h4>;
+}
+
+export default LatestOrderStatus;
+
+```
+
+Yes, as simple as this we could implement a real time communication between our client and the server. However, there are few lessons I have learned when I tried to implement a subscription for one of our recent projects.
+
+## Troubleshooting
+
+1- The subscription response must be optional. AppSync allows us to successfully define the schema with a required subscription response, but when we hook the the subscription to our frontend client, it throw an error that subscription responses must be optional.
+
+This will result to failure and error in subscription:
+
+```
+type Subscription {
+    getOrderStatus($orderId: ID!): Order! @aws_subscribe(mutations: ["updateOrderStatus"])
+}
+```
+
+2- A subscription response contains only the predicted fields which are returned by the mutation, all the other fields will be null. For instance, if we have a mutation which returns the `orderStatus` only and not the entire order object:
+
+```
+type Mutation {
+  updateOrderStatus(orderId: ID!, orderStatus: OrderStatus!): OrderStatus!
+}
+```
+
+The following subscription will send `null` for all requested data except `OrderStatus`.
+
+```
+export const GET_ORDER_STATUS = gql`
+  subscription GetOrderStatus($orderId: ID) {
+    getOrderStatus(orderId: $orderId) {
+      id
+      orderItems {
+        item
+        quantity
+      }
+      OrderStatus
+      createdAt
+    }
+  }
+`;
+```
+
+data:
+
+```
+result:{
+    data:{
+      getOrderStatus:{
+        OrderStatus: 'IN_PROGRESS',
+        orderItems: null, // orderItems is null because it was not requested in the mutation
+        createdAt: null   // createdAt is null because it was not requested in the mutation
+      }
+    }
+  }
+
+```
+
+3- Subscription arguments are indeed filters that specify when the server should notify the client.
+
+The way we have defined our subscription means that the client is interested in updates on order status when a specific orderId argument is passed through.
```

Comment:
interested in updates on order status
should be
interested in updates on an order status